### PR TITLE
More vec_bcd_ppc.h. Added:

### PR DESCRIPTION
**Pull request by munroesj52**
Signed BCD to vector __int128 conversion.
Extend/carry support for BCD add/sub.
BCD Multiply low/High combined (62-digit).
BCD Digit shifts.
Latentcy estimates for all implemented so far
Add Doxygen to explain why extend/carry got so complicated.

	* src/vec_bcd_ppc.h: Update Doxygen description.
	(bcd128_extended_0_0): : New section.
	(_BCD_CONST_PLUS_NINES, _BCD_CONST_PLUS_ONE,
	_BCD_CONST_MINUS_ONE, _BCD_CONST_ZERO, _BCD_CONST_SIGN_MASK):
	Define constants.
	(/@cond INTERNAL): Forward function declares. Add vec_bcdcpsgn,
	vec_bcdctuq, vec_bcdsrqi, vec_bcdus, vec_setbool_bcdsq,
`	vec_signbit_bcdsq.
	(vec_unpack_Decimal128, vec_BCD2DFP, vec_DFP2BCD, vec_bcdadd):
	Add Latency documentation.
	(vec_bcdaddcsq, vec_bcdaddecsq, vec_bcdaddesqm, vec_bcdcpsgn,
	vec_bcdctsq): New functions.
	(vec_bcdctub, vec_bcdctuh, vec_bcdctuw, vec_bcdctud,
	vec_bcdctuq, vec_bcddiv): Add Latency documentation.
	(vec_bcdmul): Update implementation, correct for multiply low.
	(vec_bcdmulh, vec_bcds, vec_bcdslqi, vec_bcdsluqi, vec_bcdsrqi,
	vec_bcdsruqi): New functions.
	(vec_bcdsub): Add Latency documentation.
	(vec_bcdsubcsq, vec_bcdsubecsq, vec_bcdsubesqm, vec_bcdus):
	New functions.
	(vec_cbcdaddcsq, vec_cbcdaddecsq, vec_cbcdmul): New functions.
	(vec_pack_Decimal128, vec_quantize0_Decimal128,
	vec_rdxcfzt100b, vec_rdxct100b, vec_rdxct10kh, vec_rdxct100mw,
	vec_rdxct10E16d): Add Latency documentation.
	(vec_setbool_bcdinv, vec_setbool_bcdsq, vec_signbit_bcdsq):
	New functions.
	(vec_unpack_Decimal128, vec_zndctuq):
	Add Latency documentation.

Signed-off-by: Steven Munroe <munroesj52@gmail.com>

### CODE REVIEW

File: src/vec_bcd_ppc.h [hunk 1]
```diff
@@ -554,9 +1400,478 @@ vec_bcdsub (vBCD_t a, vBCD_t b)
...
+ * Two Signed 31 digit BCD values are added, and the carry-out
+ * (the high order 32nd digit) of the sum is generated.
+ * Botht the sum and the carry are returned.
+ *
+ * |processor|Latency|Throughput|
+ * |--------:|:-----:|:---------|
+ * |power8   | 15-24 | 1/cycle  |
+ * |power9   | 6-15  | 2/cycle  |
+ *
+ * @param *cout a pointer to a 128-bit vector to recieve the BCD carry-out.
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: nit: s/*cout/cout/
cout is the "pointer", not "*cout".
There are a couple more like this, below.

File: src/vec_bcd_ppc.h [hunk 2]
```diff
@@ -554,9 +1400,478 @@ vec_bcdsub (vBCD_t a, vBCD_t b)
...
+      vBCD_t tensc = vec_bcdcpsgn (_BCD_CONST_PLUS_ONE, a);
+      c = vec_bcdcpsgn (_BCD_CONST_PLUS_ONE, t);
+      t = vec_bcdaddesqm (nines, t, tensc);
+    }
+#endif
+  *cout = c;
+  return (t);
+}
+
+/** \brief Combined Decimal Add Extended & write Carry Signed Quadword.
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/write/Write/

File: src/vec_bcd_ppc.h [hunk 3]
```diff
@@ -554,9 +1400,478 @@ vec_bcdsub (vBCD_t a, vBCD_t b)
...
+ * Two Signed 31 digit values and a signed carry-in are added together
+ * and the carry-out (the high order 32nd digit) of the sum is
+ * returned.
+ *
+ * |processor|Latency|Throughput|
+ * |--------:|:-----:|:---------|
+ * |power8   | 54-63 | 1/cycle  |
+ * |power9   | 15-24 | 2/cycle  |
+ *
+ * @param *cout a pointer to a 128-bit vector to recieve the BCD carry-out.
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/*cout/cout/

File: src/vec_bcd_ppc.h [hunk 4]
```diff
@@ -554,9 +1400,478 @@ vec_bcdsub (vBCD_t a, vBCD_t b)
...
+ * Then DFP Decimal shift will isolate 32nd digit.
+ *
+ * |processor|Latency|Throughput|
+ * |--------:|:-----:|:---------|
+ * |power8   |107-413| 1/cycle  |
+ * |power9   |115-294| 1/cycle  |
+ *
+ * @param a a 128-bit vector treated as a signed BCD 31 digit value.
+ * @param b a 128-bit vector treated as a signed BCD 31 digit value.
+ * @param *p_high pointer 128-bit vector to receive the high 31-digits
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/*p_high/p_high/

File: src/vec_bcd_ppc.h [hunk 5]
```diff
@@ -880,9 +2230,171 @@ vec_rdxct10e32q (vui64_t vra)
...
+ *  this implementation only tests for a valid minus sign nibble.
+ *  Otherwise the BCD value is assumed to be positive.
+ *
+ *  |processor| Latency |Throughput|
+ *  |--------:|:-------:|:---------|
+ *  |power8   | 15 - 26 | 2/cycle  |
+ *  |power9   | 5 - 14  | 2/cycle  |
+ *
+ *  @param vra a 128-bit vector treated as signed BCD quadword.
+ *  @return a none-zero value if the sign is negative.
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/none-zero/non-zero/

File: src/vec_bcd_ppc.h [hunk 6]
```diff
@@ -880,9 +2230,171 @@ vec_rdxct10e32q (vui64_t vra)
...
+static inline int
+vec_signbit_bcdsq (vBCD_t vra)
+{
+  int result;
+#if defined (_ARCH_PWR8) && (__GNUC__ > 6)
+  result = __builtin_bcdsub_lt ((vi128_t) vra, (vi128_t) _BCD_CONST_ZERO, 0);
+#else
+  const vui32_t sign_mask = vec_splat_u32(15);
+  const vui32_t minus_sign = (vui32_t) CONST_VINT128_W(0x0b, 0x0d, 0x0b, 0x0d);
+//  const vui32_t plus_sign = (vui32_t) CONST_VINT128_W(0x0a, 0x0c, 0x0e, 0x0f);
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: did you want to keep this commented-out line?

File: src/vec_bcd_ppc.h [hunk 7]
```diff
@@ -216,6 +216,251 @@
...
+ * the bcdadd/sub instructions. But bcdadd/sub instructions do post
+ * status to CR field 6 which includes:
+ * - Result is less than zero (CR.bit[56])
+ * - Result is greater than zero (CR.bit[57])
+ * - Result is equal to zero (CR.bit[58])
+ * - Result overflowed (CR.bit[59])
+ *
+ * which provides a basis for BCD comparison and the overflow may be
+ * used for carry/extend logic. The GCC compiler provides built-ins to
+ * generated the bcdadd/sub and test the resulting CR bits in if
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 8]
```diff
@@ -216,6 +216,251 @@
...
+ * statements.
+ *
+ * Unfortunately the Overflow flag generated by bcdadd/bcdsub, is not
+ * a true carry/borrow. If the operands have the same sign for bcdadd
+ * (different sign for bcdsub)
+ * and there is a carry out of the high order digit, then:
+ * - The sum is truncated to the low order 31 digits
+ * - The sign code matches the operand signs
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: How about:
"The sum's sign matches the operands signs"

File: src/vec_bcd_ppc.h [hunk 9]
```diff
@@ -216,6 +216,251 @@
...
+ * - Result is greater than zero (CR.bit[57])
+ * - Result is equal to zero (CR.bit[58])
+ * - Result overflowed (CR.bit[59])
+ *
+ * which provides a basis for BCD comparison and the overflow may be
+ * used for carry/extend logic. The GCC compiler provides built-ins to
+ * generated the bcdadd/sub and test the resulting CR bits in if
+ * statements.
+ *
+ * Unfortunately the Overflow flag generated by bcdadd/bcdsub, is not
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 10]
```diff
@@ -216,6 +216,251 @@
...
+ * and there is a carry out of the high order digit, then:
+ * - The sum is truncated to the low order 31 digits
+ * - The sign code matches the operand signs
+ * - The overflow flag (CR.bit[59]) is set.
+ *
+ * \note overflow is only set in conjunction with greater than zero
+ * (positive) or less than zero (negative) results. This implies that
+ * BCD carries are tri-state; +1, 0, or -1.
+ *
+ * This can be used to simulate a <B>Add and write-Carry</B> operation.
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 11]
```diff
@@ -216,6 +216,251 @@
...
+ * However if the operands have different signs the bcdadd
+ * (same sign for bcdsub) the operation does the following:
+ * - The smaller magnitude is subtracted from the larger magnitude.
+ * - The sign code matches the sign of the larger magnitude.
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 12]
```diff
@@ -216,6 +216,251 @@
...
+ * - The ox_flag is NOT set.
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 13]
```diff
@@ -216,6 +216,251 @@
...
+ *
+ * For a simple BCD add this is the desired result (overflow is
+ * avoided and the borrow is recorded in the sign).
+ * But for multiple precision BCD operation, this will delay
+ * propagation of borrows to the higher order digits and the result a
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 14]
```diff
@@ -216,6 +216,251 @@
...
+ * simplify the code. The simplified multiple precision BCD use case
+ * looks like this:
+ * \code
+  // r_h|r_l = a_h|a_l + b_h|b_l
+  r_l = vec_bcdadd (a_l, b_l);
+  c_l = vec_bcdaddcsq (a_l, b_l);
+  r_h = vec_bcdaddesqm (a_h, b_h, c_l)
+ * \endcode
+ *
+ * But we should look at some more example before we assume we have a
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 15]
```diff
@@ -216,6 +216,251 @@
...
+ *   21000000000000000000000000000008
+ * - 19000000000000000000000000000008
+ * = 02000000000000000000000000000000
+ * \endcode
+ * The multiple precision BCD would look like this:
+ * \code
+ *   0000000000000000000000000000002c 1000000000000000000000000000008c
+ * + 0000000000000000000000000000001d 9000000000000000000000000000008d
+ * \endcode
+ * But the the example coide above we do expected result:
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: But with the example code above we expected result:

File: src/vec_bcd_ppc.h [hunk 16]
```diff
@@ -352,20 +623,269 @@ vec_bcdadd (vBCD_t a, vBCD_t b)
...
+#if defined ( _ARCH_PWR8) && (__GNUC__ > 6)
+  t = _BCD_CONST_ZERO;
+  if (__builtin_expect (__builtin_bcdadd_ov ((vi128_t) a, (vi128_t) b, 0), 0))
+    {
+#ifdef _ARCH_PWR9
+      t = (vBCD_t) __builtin_bcdadd ((vi128_t) a, (vi128_t) b, 0);
+      t = vec_bcdcpsgn (_BCD_CONST_PLUS_ONE, t);
+#else
+      if (__builtin_bcdadd_gt ((vi128_t) a, (vi128_t) b, 0))
+      t = _BCD_CONST_PLUS_ONE;
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 17]
```diff
@@ -352,20 +623,269 @@ vec_bcdadd (vBCD_t a, vBCD_t b)
...
+      else
+      t = _BCD_CONST_MINUS_ONE;
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 18]
```diff
@@ -545,6 +1390,7 @@ vec_bcdsub (vBCD_t a, vBCD_t b)
   t = (vBCD_t) __builtin_bcdsub ((vi128_t) a, (vi128_t) b, 0);
 #endif
 #else
+  const vui32_t mz = CONST_VINT128_W (0, 0, 0, 0x0000000d);
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: missing fix:

  // fix up spurious negative zeros
  if (vec_all_eq((vui32_t) t, mz))
    t = _BCD_CONST_ZERO;

File: src/vec_bcd_ppc.h [hunk 19]
```diff
@@ -554,9 +1400,478 @@ vec_bcdsub (vBCD_t a, vBCD_t b)
...
+  vrt = (vBCD_t) t;
+#endif
+  return (vrt);
+}
+
+/** \brief Combined Decimal Add & write Carry Signed Quadword.
+ *
+ * Two Signed 31 digit BCD values are added, and the carry-out
+ * (the high order 32nd digit) of the sum is generated.
+ * Botht the sum and the carry are returned.
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 1 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 2 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: Ok
But you talked to Brett and convinced him that the PowerISA is wrong about write-Carry and needs to change, right?

File: src/vec_bcd_ppc.h (hunk 3 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: yea yea OK

File: src/vec_bcd_ppc.h (hunk 4 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: yup OK

File: src/vec_bcd_ppc.h (hunk 5 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 6 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: Removed

**Comment by munroesj52** [comment]
Updated per comments and commited.

**Pull request merged**